tasks/managed/filter-published-fbc-images/filter-published-fbc-images.yaml

```
@@ -0,0 +1,453 @@
+        set -eux
+        
+        echo "Extracting Pyxis credentials from mounted secret"
+        PYXIS_CERT="$(cat /etc/secrets/cert)"
+        PYXIS_KEY="$(cat /etc/secrets/key)"
+        echo "${PYXIS_CERT}" > /var/workdir/crt
+        echo "${PYXIS_KEY}" > /var/workdir/key
+    - name: filter-already-released-images
+      image: quay.io/konflux-ci/release-service-utils:cee1fd0191e750b0736d8197701d91ce2f308e03
+      computeResources:
+        limits:
+          memory: 512Mi
+        requests:
+          memory: 512Mi
+          cpu: 350m
+      script: |
+        #!/usr/bin/env bash
+        set -eux
+
+        SNAPSHOT_FILE="$(params.dataDir)/$(params.snapshotPath)"
+        DATA_FILE="$(params.dataDir)/$(params.dataPath)"
+        FILTERED_SNAPSHOT_FILE="$(params.dataDir)/filtered-snapshot.json"
+
+        if [ ! -f "${SNAPSHOT_FILE}" ]; then
+            echo "ERROR: No valid snapshot file at ${SNAPSHOT_FILE}"
+            exit 1
+        fi
+
+        if [ ! -f "${DATA_FILE}" ]; then
+            echo "ERROR: No valid data file at ${DATA_FILE}"
+            exit 1
+        fi
+
+        echo "==============================================================="
+        echo "Filtering Already-Released FBC Fragments via Pyxis Index Query"
+        echo "==============================================================="
+        echo ""
+
+        pyxis_server=$(jq -r '.pyxis.server // "production"' "${DATA_FILE}")
+        target_index=$(jq -r '.fbc.targetIndex // ""' "${DATA_FILE}")
+        
+        if [ -z "$target_index" ]; then
+          echo "ERROR: fbc.targetIndex not found in data file"
+          echo "This task requires targetIndex to query for existing index images"
+          exit 1
+        fi
+        
+        if [[ "$target_index" == *"{{ OCP_VERSION }}"* ]]; then
+          echo "Target index contains {{ OCP_VERSION }} template, resolving..."
+          
+          container_image=$(jq -r '.components[0].containerImage' "${SNAPSHOT_FILE}")
```

Hardcoded `component[0]`? That seems wrong. I'm pretty sure we support multicomponent fbc releases now.

---

Or is this because all the components will always use the same OCP version anyway, so it's enough to just check one?

---

The code assumes all components in a multicomponent fbc release use the same OCP version, which is why it only checks components[0].
There's only one targetIndex in the RPA data with a single {{ OCP_VERSION }} placeholder
All fbc fragments from a release get published to the same index image
Looking at our test configs, component2.yaml even has a comment: context: "4.13" # Same as component1 for batching
But definitely we could add such validation: instead of silently assuming, we should verify that all components use the same OCP version and fail with a clear error if they don't.
WDYT?

---

@mmalina implemented as described

---

OK, so it says "same component for batching". Does this imply that in theory they could be different? It sounds like it, no? @theflockers can you comment?

---

it's a hard requirement, not just for batching

why they MUST be the same:
- all components in a multicomponent FBC release are published to one targetIndex
- each targetIndex is tied to a specific OCP version:
          - catalog:v4.13 ← only v4.13 fragments
          - catalog:v4.14 ← only v4.14 fragments

For example: IIB builds fail if you try to add a v4.14 fragment to a v4.13 index; IIB will reject it

This particular comment is not my code; it was created by someone else earlier. 
The word "batching" might refer to how IIB processes multiple fragments in one request, but the exact OCP version requirement isn't for batching—the underlying architecture requires it; but I cannot be sure.
Would you like me to update that comment to be more accurate?
I can change the comment to something like this: "MUST match component1 OCP version (single targetIndex requirement)".

---

I would like @theflockers to comment. He is likely the owner of that comment. You may be right. In theory we could allow multiple components to have different targets - and then each would have a separate IIB request. And the publishing would also be separate. But we probably don't support that - each fbc release probably has to use exactly one target version.

---

The pipeline can be used the following way:

1. Multiple snapshots of the same application, different OCP versions
2. Multiple snapshots of different applications, same OCP version.

So, if you follow in the first case, indeed. `component[0]` would be fine, because you are handling different versions of the same component, but not the second, when the application might be another one, hence you could be potentially making a global decision based in a single application.

---

About the "batching": I can 't explain much about how this works. I understood that is something that is supported by IIB then the code was changed to support that as well. IMHO it is harder to read the code after the change - I know Andrew had some AI assisted code by Claude - so maybe that is why I struggle to understand what is happening there now.

---

> Multiple snapshots of the same application

what do you mean by this? I don't follow. a release always has exactly one snaphot...

---

I think batching is just that you add multiple fragments in one iib request. But an iib request is always for a specific openshift version I think.

---

@theflockers I appreciate your detailed and informative feedback—it's been extremely helpful in clarifying many points!

---

Gentlemen, before I move forward, could you please confirm whether any modifications or fixes are required for my current code, or if the existing solution meets your requirements?

---

Sorry @mmalina I wanted to say multiple applications in the same snapshot, not multiple snapshots.

---

> I wanted to say multiple applications in the same snapshot, not multiple snapshots.

ok, no problem, but I don't understand this either :) I don't think you can have multiple applications in one snapshot. just multiple components of the same applications. or is this some other meaning of application?

---

@elenagerman You need to change to look the other components. Because the assumption is that you can have N applications in the same snapshot. Eventually you might have an fbc_fragment that was already published for that version inside the snapshot, I suppose you want to handle this case as well. 
So IMHO the best would be check all components of the snapshot instead of just the index 0.

---

> Gentlemen, before I move forward, could you please confirm whether any modifications or fixes are required for my current code, or if the existing solution meets your requirements?

I think it should be fine to only do the filtering for production builds. But it should be done for each target index, right?

---

@theflockers 
We already check all components!
The filtering logic (lines 403-426) loops through every component in the snapshot:

- Extracts each fragment's digest individually
- Checks each against Pyxis published fragments
- Keeps only unpublished ones

This handles partial publishing scenarios (e.g., 2 out of 3 fragments already published).
The components[0] usage you see is only for OCP version extraction. We then validate that all other components match that version (lines 197-238). Since all fragments must target the same targetIndex, they must have the same OCP version anyway.
So yes - N applications, individual filtering per fragment, already implemented.

---

@mmalina 
Great point! You're right - filtering should be per-targetIndex. 
Currently, we enforce single OCP version (all components → same targetIndex), so one Pyxis query works. But if we support multi-OCP releases (like PR #1565 is moving toward), I'd need to change solution logic again:

- Group components by resolved targetIndex
- Query Pyxis for each targetIndex separately
- Filter each group independently

Note: probably we need to disscuss it on the next standup meeting, it is not a good practice when two engineers working on the same part of code simultaniously.

---

@mmalina code was modified to support multiple releases; the [doc](https://docs.google.com/document/d/1K-pRERd1Y4zNrE-8WjDHorUTP6DPTCX_OoxZeLUsB3g/edit?tab=t.0) was updated accordingly. 
Thank you!